[Feat] 패키지 업데이트 및 린터 설정

seokzin opened the pull request:
## 📌 이슈
- close #108
- close #139
- close #141

<br/>

## 📝 설명

### ✔️ ESLint
#### 1. 린트 적용 및 오류 수정 완료
오류 수정 과정에서 **리팩토링**을 진행했는데 논의가 필요해 보여요.

- 2ecd1cd960e3d961996b088f0131f79d5d288d73
- d81d64c9d4e6d0e3859524f90c4708ebecfc8c3d 
- a3258a5312ec3474d4d1faccbc0bd1582e9a9cf9

#### 2. Import/Export 정렬
가나다순까지 정렬되는 꽤 야무진 플러그인이네요! 👍

#### 3. Next 린트 범위
- next lint 명령은 기본적으로 `/pages`, `/components/` `/lib`만 해당돼요.
- 그래서 `--dir`를 통해 범위를 전체로 변경했어요.

<br />

### ✔️ stylelint
#### 1. 린트 적용 및 postcss 버전 의존성 문제 해결
```
TypeError: Cannot read properties of undefined (reading 'stringify')
```
- 다음 에러가 발생했는데 postcss 버전 의존성 문제로 추정돼요. 
- https://github.com/stylelint-scss/stylelint-config-standard-scss/issues/5#issuecomment-961709882

#### 2. 스크립트 주의 사항
- 스크립트에서 곁따옴표와 홑따옴표를 혼용하면 에러가 발생해요.
- [이스케이프 시퀀스](https://stackoverflow.com/questions/54543063/how-can-i-suppress-the-no-files-matching-the-pattern-message-in-eslint)를 통해 해결했어요. 

#### 4. 일부 규칙들 논의
- [alpha-value-notation](https://stylelint.io/user-guide/rules/list/alpha-value-notation)
- [color-function-notation](https://stylelint.io/user-guide/rules/list/color-function-notation)

<br />

### ✔️  패키지 업데이트
Yarn으로 [일괄 업데이트](https://stackoverflow.com/questions/62650640/yarn-how-do-i-update-each-dependency-in-package-json-to-the-latest-version/62650754#62650754) 했으니 재설치 해주세요.
새 환경에서 테스트하다가 스크립트 이슈를 발견했어요. (28ffe1e0e547960c819af40fe56e30836f8e7bd6)

<br/>

### ✔️ Git Hooks
- husky + lint-staged를 통해 린트 및 테스트를 자동화했어요.
  - pre-commit : lint + prettier + stylelint
  - pre-push ; jest --coverage

<br/>

## 📷 스크린샷

![image](https://user-images.githubusercontent.com/43740455/188553044-48ff2686-a194-4e9e-a5dc-d48ad6045bb3.png)

깨끗 🤗

seokzin committed:
chore: Delete   prettier/prettier 에러 무시

seokzin committed:
docs: PR 템플릿 여백 추가

seokzin committed:
chore: 'React' must be in scope when using JSX 에러 무시

seokzin committed:
refac: 프리티어 수동 적용

seokzin committed:
feat: 패키지 일괄 업데이트

seokzin committed:
chore: Prettier 설정

- 이외 설정들은 기본값과 중복되어서 제거

seokzin committed:
feat: app 프로퍼티 타입 롤백

- #055a9d1
- 이미 함수형 컴포넌트에 수정을 가해 린트 에러 발생
- 무의미한 수정이라 생각하여 초기 상태로 롤백

seokzin committed:
refac: any 타입 제거

- HTMLInputElement의 기본 타입 (value: string) 참고

seokzin committed:
refac: 인터페이스 간소화

seokzin committed:
fix: children 프로퍼티를 합성 컴포넌트로 대체

- no-children-prop 규칙

seokzin committed:
fix: 올바른 타입으로 변경

seokzin committed:
fix: 타입 변경에 따른 입력값 변경

seokzin committed:
feat: 로그인 상태 의존성 위임

- 객체지향 원칙을 토대로 불필요한 의존성을 위임했어요

seokzin committed:
fix: 더미 함수 변경

- no-empty-function 오류 해결 위한 일종의 더미

seokzin committed:
refac: 타입 단언을 통해 any 타입 제거

- includes는 집합 관계를 엄격히 따져요.
  (https://github.com/Microsoft/TypeScript/issues/26255)

- unknown을 통한 이중 타입 단언으로 타입을 조심스래 맞춰야 해요.
  (이펙티브 타입스크립트 item 9)

- 그리고 size 또한 undefined가 아님을 보장해야 돼요.

seokzin committed:
refac: Prettier 적용

seokzin committed:
chore: 린터 플러그인 추가

seokzin committed:
refac: import/export 정렬

seokzin committed:
chore: 스크립트 '&&' 추가

seokzin committed:
refac: 인터페이스 통합

devpla commented on components/Text/Text.styled.ts:
```diff
@@ -1,26 +1,28 @@
-  letter-spacing: ${({ size }: any) =>
-    TitleList.includes(size) ? "-0.03rem" : "-0.02rem"};
+  letter-spacing: ${({ size }) =>
+    size && (TitleList as unknown as FontSizeKey).includes(size)
```
`font-size: ${({ size = 'body1', theme }) => theme.fontSize[size]}px;`
위에 이 코드가 있어서, 여기서도 같은 방식으로 처리하면 좋을 것 같아요!

devpla commented on components/Navbar/Navbar.types.ts:
```diff
@@ -1,34 +1,35 @@
-    NavOptions {}
+    NavOptions {
+  topItemsLogin: itemOptions[]
+  topItemsLogout: itemOptions[]
```
NavOptions를 따로 분리하는게 불필요하다고 생각되는데요.
다른 컴포넌트도 Options를 다 분리했던걸로 기억하는데, 요거 다음에 합치면 좋을 것 같아요

devpla commented on types/ComponentProps.ts:
```diff
@@ -1,21 +1,21 @@
-export interface ChildrenProps<Children = React.ReactNode> {
-  children?: Children;
+export interface ChildrenProps {
+  children?: React.ReactNode
```
굿굿쟙 👍

seokzin committed:
feat: stylelint 설정

- TypeError: Cannot read properties of undefined (reading 'stringify')
- post-css 버전 의존성 이슈
  (https://github.com/stylelint-scss/stylelint-config-standard-scss/issues/5#issuecomment-961709882)

seokzin committed:
refac: Prettier 적용

seokzin committed:
feat: a11y 린터 추가

seokzin committed:
chore: stylelint 프로시저 제거

seokzin committed:
feat: stylelint 규칙 선언 및 적용

seokzin committed:
chore: 린터 범위 모든 파일로 확대

seokzin committed:
refac: 린터 적용

seokzin committed:
feat: husky 설치 및 언어 통계 제외

seokzin committed:
feat: husky lint-staged 설치 및 적용

seokzin commented:
사용하면서 불편함을 찾아 보완할 예정이에요.
특히 pre-push의 테스트 자동화는 pre-commit과 합치고 싶어요.
